File: sdk/nodejs/runtime/callbacks.ts
// Copyright 2016-2024, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as grpc from "@grpc/grpc-js";
import { randomUUID } from "crypto";
import * as jspb from "google-protobuf";
import * as gstruct from "google-protobuf/google/protobuf/struct_pb";
import * as log from "../log";
import { output } from "../output";
import * as callrpc from "../proto/callback_grpc_pb";
import * as callproto from "../proto/callback_pb";
import { Callback, CallbackInvokeRequest, CallbackInvokeResponse } from "../proto/callback_pb";
import * as resrpc from "../proto/resource_grpc_pb";
import * as resproto from "../proto/resource_pb";
import {
    Alias,
    ComponentResourceOptions,
    CustomResourceOptions,
    DependencyProviderResource,
    DependencyResource,
    ProviderResource,
    Resource,
    ResourceHook,
    ResourceOptions,
    ResourceTransform,
    ResourceTransformArgs,
    URN,
    rootStackResource,
} from "../resource";
import { InvokeOptions, InvokeTransform, InvokeTransformArgs } from "../invoke";

import { hookBindingFromProto, mapAliasesForRequest, prepareHooks } from "./resource";
import { deserializeProperties, serializeProperties, unknownValue } from "./rpc";
import { debuggablePromise } from "./debuggable";
import { rpcKeepAlive } from "./settings";

/**
 * Raises the gRPC Max Message size from `4194304` (4mb) to `419430400` (400mb)
 *
 * @internal
 */
const maxRPCMessageSize: number = 1024 * 1024 * 400;

type CallbackFunction = (args: Uint8Array) => Promise<jspb.Message>;

export interface ICallbackServer {
    registerTransform(callback: ResourceTransform): Promise<callproto.Callback>;
    registerStackTransform(callback: ResourceTransform): void;
    registerStackInvokeTransform(callback: InvokeTransform): void;
    registerStackInvokeTransformAsync(callback: InvokeTransform): Promise<callproto.Callback>;
    registerResourceHook(hook: ResourceHook): Promise<void>;
    shutdown(): void;
    // Wait for any pendind registerStackTransform calls to complete.
    awaitStackRegistrations(): Promise<void>;
}

export class CallbackServer implements ICallbackServer {
    private readonly _callbacks = new Map<string, CallbackFunction>();
    private readonly _monitor: resrpc.IResourceMonitorClient;
    private readonly _server: grpc.Server;
    private readonly _target: Promise<string>;
    private _pendingRegistrations: number = 0;
    private _awaitQueue: ((reason?: any) => void)[] = [];

    constructor(monitor: resrpc.IResourceMonitorClient) {
        this._monitor = monitor;

        this._server = new grpc.Server({
            "grpc.max_receive_message_length": maxRPCMessageSize,
        });

        const implementation: callrpc.ICallbacksServer = {
            invoke: this.invoke.bind(this),
        };
        this._server.addService(callrpc.CallbacksService, implementation);

        const self = this;
        this._target = new Promise<string>((resolve, reject) => {
            self._server.bindAsync(`127.0.0.1:0`, grpc.ServerCredentials.createInsecure(), (err, port) => {
                if (err !== null) {
                    reject(err);
                    return;
                }

                // The server takes a while to _actually_ startup so we need to keep trying to send an invoke
                // to ourselves before we resolve the address to tell the engine about it.
                const target = `127.0.0.1:${port}`;

                const client = new callrpc.CallbacksClient(target, grpc.credentials.createInsecure());

                const connect = () => {
                    client.invoke(new CallbackInvokeRequest(), (error, _) => {
                        if (error?.code === grpc.status.UNAVAILABLE) {
                            setTimeout(connect, 1000);
                            return;
                        }
                        // The expected error given we didn't give a token to the invoke.
                        if (error?.details === "callback not found: ") {
                            resolve(target);
                            return;
                        }
                        reject(error);
                    });
                };
                connect();
            });
        });
    }

    awaitStackRegistrations(): Promise<void> {
        if (this._pendingRegistrations === 0) {
            return Promise.resolve();
        }
        return new Promise<void>((resolve, reject) => {
            this._awaitQueue.push((reason?: any) => {
                if (reason !== undefined) {
                    reject(reason);
                } else {
                    resolve();
                }
            });
        });
    }

    shutdown(): void {
        this._server.forceShutdown();
    }

    private async invoke(
        call: grpc.ServerUnaryCall<CallbackInvokeRequest, CallbackInvokeResponse>,
        callback: grpc.sendUnaryData<CallbackInvokeResponse>,
    ) {
        const req = call.request;

        const cb = this._callbacks.get(req.getToken());
        if (cb === undefined) {
            const err = new grpc.StatusBuilder();
            err.withCode(grpc.status.INVALID_ARGUMENT);
            err.withDetails("callback not found: " + req.getToken());
            callback(err.build());
            return;
        }

        try {
            const response = await cb(req.getRequest_asU8());
            const resp = new CallbackInvokeResponse();
            resp.setResponse(response.serializeBinary());
            callback(null, resp);
        } catch (e) {
            const err = new grpc.StatusBuilder();
            err.withCode(grpc.status.UNKNOWN);
            if (e instanceof Error) {
                err.withDetails(e.message);
            } else {
                err.withDetails(JSON.stringify(e));
            }
            callback(err.build());
        }
    }

    async registerTransform(transform: ResourceTransform): Promise<callproto.Callback> {
        const cb = async (bytes: Uint8Array): Promise<jspb.Message> => {
            const request = resproto.TransformRequest.deserializeBinary(bytes);

            let opts = request.getOptions() || new resproto.TransformResourceOptions();

            let ropts: ResourceOptions;
            if (request.getCustom()) {
                ropts = {
                    deleteBeforeReplace: opts.getDeleteBeforeReplace(),
                    additionalSecretOutputs: opts.getAdditionalSecretOutputsList(),
                } as CustomResourceOptions;
            } else {
                const providers: Record<string, ProviderResource> = {};
                for (const [key, value] of opts.getProvidersMap().entries()) {
                    providers[key] = new DependencyProviderResource(value);
                }
                ropts = {
                    providers: providers,
                } as ComponentResourceOptions;
            }

            ropts.aliases = opts.getAliasesList().map((alias): string | Alias => {
                if (alias.hasUrn()) {
                    return alias.getUrn();
                } else {
                    const spec = alias.getSpec();
                    if (spec === undefined) {
                        throw new Error("alias must have either a urn or a spec");
                    }
                    const nodeAlias: Alias = {
                        name: spec.getName(),
                        type: spec.getType(),
                        project: spec.getProject(),
                        stack: spec.getStack(),
                        parent: spec.getParenturn() !== "" ? new DependencyResource(spec.getParenturn()) : undefined,
                    };

                    if (spec.getNoparent()) {
                        nodeAlias.parent = rootStackResource;
                    }

                    return nodeAlias;
                }
            });
            const timeouts = opts.getCustomTimeouts();
            if (timeouts !== undefined) {
                ropts.customTimeouts = {
                    create: timeouts.getCreate(),
                    update: timeouts.getUpdate(),
                    delete: timeouts.getDelete(),
                };
            }
            ropts.hooks = hookBindingFromProto(opts.getHooks());
            ropts.deletedWith =
                opts.getDeletedWith() !== "" ? new DependencyResource(opts.getDeletedWith()) : undefined;
            ropts.dependsOn = opts.getDependsOnList().map((dep) => new DependencyResource(dep));
            ropts.ignoreChanges = opts.getIgnoreChangesList();
            ropts.parent = request.getParent() !== "" ? new DependencyResource(request.getParent()) : undefined;
            ropts.pluginDownloadURL = opts.getPluginDownloadUrl() !== "" ? opts.getPluginDownloadUrl() : undefined;
            ropts.protect = opts.getProtect();
            ropts.provider = opts.getProvider() !== "" ? new DependencyProviderResource(opts.getProvider()) : undefined;
            ropts.replaceOnChanges = opts.getReplaceOnChangesList();
            ropts.retainOnDelete = opts.getRetainOnDelete();
            ropts.version = opts.getVersion() !== "" ? opts.getVersion() : undefined;

            const props = request.getProperties();

            const args: ResourceTransformArgs = {
                custom: request.getCustom(),
                type: request.getType(),
                name: request.getName(),
                props: props === undefined ? {} : deserializeProperties(props),
                opts: ropts,
            };

            const result = await transform(args);

            const response = new resproto.TransformResponse();
            if (result === undefined) {
                response.setProperties(request.getProperties());
                response.setOptions(request.getOptions());
            } else {
                const mprops = await serializeProperties("props", result.props);
                response.setProperties(gstruct.Struct.fromJavaScript(mprops));

                // Copy the options over.
                if (result.opts !== undefined) {
                    opts = new resproto.TransformResourceOptions();

                    if (result.opts.aliases !== undefined) {
                        const aliases = [];
                        const uniqueAliases = new Set<Alias | URN>();
                        for (const alias of result.opts.aliases || []) {
                            const aliasVal = await output(alias).promise();
                            if (!uniqueAliases.has(aliasVal)) {
                                uniqueAliases.add(aliasVal);
                                aliases.push(aliasVal);
                            }
                        }

                        opts.setAliasesList(await mapAliasesForRequest(aliases, request.getParent()));
                    }
                    if (result.opts.customTimeouts !== undefined) {
                        const customTimeouts = new resproto.RegisterResourceRequest.CustomTimeouts();
                        if (result.opts.customTimeouts.create !== undefined) {
                            customTimeouts.setCreate(result.opts.customTimeouts.create);
                        }
                        if (result.opts.customTimeouts.update !== undefined) {
                            customTimeouts.setUpdate(result.opts.customTimeouts.update);
                        }
                        if (result.opts.customTimeouts.delete !== undefined) {
                            customTimeouts.setDelete(result.opts.customTimeouts.delete);
                        }
                        opts.setCustomTimeouts(customTimeouts);
                    }
                    if (result.opts.deletedWith !== undefined) {
                        opts.setDeletedWith(await result.opts.deletedWith.urn.promise());
                    }
                    if (result.opts.dependsOn !== undefined) {
                        const resolvedDeps = await output(result.opts.dependsOn).promise();
                        const deps = [];
                        if (Resource.isInstance(resolvedDeps)) {
                            deps.push(await resolvedDeps.urn.promise());
                        } else {
                            for (const dep of resolvedDeps) {
                                deps.push(await dep.urn.promise());
                            }
                        }
                        opts.setDependsOnList(deps);
                    }
                    if (result.opts.ignoreChanges !== undefined) {
                        opts.setIgnoreChangesList(result.opts.ignoreChanges);
                    }
                    if (result.opts.pluginDownloadURL !== undefined) {
                        opts.setPluginDownloadUrl(result.opts.pluginDownloadURL);
                    }
                    if (result.opts.protect !== undefined) {
                        opts.setProtect(result.opts.protect);
                    }
                    if (result.opts.provider !== undefined) {
                        const providerURN = await result.opts.provider.urn.promise();
                        const providerID = (await result.opts.provider.id.promise()) || unknownValue;
                        opts.setProvider(`${providerURN}::${providerID}`);
                    }
                    if (result.opts.replaceOnChanges !== undefined) {
                        opts.setReplaceOnChangesList(result.opts.replaceOnChanges);
                    }
                    if (result.opts.retainOnDelete !== undefined) {
                        opts.setRetainOnDelete(result.opts.retainOnDelete);
                    }
                    if (result.opts.version !== undefined) {
                        opts.setVersion(result.opts.version);
                    }
                    if (result.opts.hooks !== undefined) {
                        opts.setHooks(await prepareHooks(result.opts.hooks, request.getName()));
                    }

                    if (request.getCustom()) {
                        const copts = result.opts as CustomResourceOptions;
                        if (copts.deleteBeforeReplace !== undefined) {
                            opts.setDeleteBeforeReplace(copts.deleteBeforeReplace);
                        }
                        if (copts.additionalSecretOutputs !== undefined) {
                            opts.setAdditionalSecretOutputsList(copts.additionalSecretOutputs);
                        }
                    } else {
                        const copts = result.opts as ComponentResourceOptions;

                        if (copts.providers !== undefined) {
                            const providers = opts.getProvidersMap();

                            if (copts.providers && !Array.isArray(copts.providers)) {
                                for (const k in copts.providers) {
                                    if (Object.prototype.hasOwnProperty.call(copts.providers, k)) {
                                        const v = copts.providers[k];
                                        if (k !== v.getPackage()) {
                                            const message = `provider resource map where key ${k} doesn't match provider ${v.getPackage()}`;
                                            log.warn(message);
                                        }
                                    }
                                }
                            }
                            const provs = Object.values(copts.providers);
                            for (const prov of provs) {
                                const providerURN = await prov.urn.promise();
                                const providerID = (await prov.id.promise()) || unknownValue;
                                providers.set(prov.getPackage(), `${providerURN}::${providerID}`);
                            }
                            opts.clearProvidersMap();
                        }
                    }
                }
                response.setOptions(opts);
            }

            return response;
        };
        const tryCb = async (bytes: Uint8Array): Promise<jspb.Message> => {
            try {
                return await cb(bytes);
            } catch (e) {
                throw new Error(`transform failed: ${e}`);
            }
        };
        const uuid = randomUUID();
        this._callbacks.set(uuid, tryCb);
        const req = new Callback();
        req.setToken(uuid);
        req.setTarget(await this._target);
        return req;
    }

    registerStackTransform(transform: ResourceTransform): void {
        this._pendingRegistrations++;

        this.registerTransform(transform)
            .then(
                (req) => {
                    return new Promise((resolve, reject) => {
                        this._monitor.registerStackTransform(req, (err, _) => {
                            if (err !== null) {
                                // Remove this from the list of callbacks given we didn't manage to actually register it.
                                this._callbacks.delete(req.getToken());
                                reject();
                            } else {
                                resolve();
                            }
                        });
                    });
                },
                (err) => log.error(`failed to register stack transform: ${err}`),
            )
            .finally(() => {
                this._pendingRegistrations--;
                if (this._pendingRegistrations === 0) {
                    const queue = this._awaitQueue;
                    this._awaitQueue = [];
                    for (const waiter of queue) {
                        waiter();
                    }
                }
            });
    }

    async registerStackInvokeTransformAsync(transform: InvokeTransform): Promise<callproto.Callback> {
        const cb = async (bytes: Uint8Array): Promise<jspb.Message> => {
            const request = resproto.TransformInvokeRequest.deserializeBinary(bytes);

            let opts = request.getOptions() || new resproto.TransformInvokeOptions();

            const ropts: InvokeOptions = {};
            ropts.pluginDownloadURL = opts.getPluginDownloadUrl() !== "" ? opts.getPluginDownloadUrl() : undefined;
            ropts.provider = opts.getProvider() !== "" ? new DependencyProviderResource(opts.getProvider()) : undefined;
            ropts.version = opts.getVersion() !== "" ? opts.getVersion() : undefined;

            const invokeArgs = request.getArgs();

            const args: InvokeTransformArgs = {
                token: request.getToken(),
                args: invokeArgs === undefined ? {} : deserializeProperties(invokeArgs),
                opts: ropts,
            };

            const result = await transform(args);

            const response = new resproto.TransformInvokeResponse();
            if (result === undefined) {
                response.setArgs(request.getArgs());
                response.setOptions(request.getOptions());
            } else {
                const margs = await serializeProperties("args", result.args);
                response.setArgs(gstruct.Struct.fromJavaScript(margs));

                // Copy the options over.
                if (result.opts !== undefined) {
                    opts = new resproto.TransformInvokeOptions();

                    if (result.opts.pluginDownloadURL !== undefined) {
                        opts.setPluginDownloadUrl(result.opts.pluginDownloadURL);
                    }
                    if (result.opts.provider !== undefined) {
                        const providerURN = await result.opts.provider.urn.promise();
                        const providerID = (await result.opts.provider.id.promise()) || unknownValue;
                        opts.setProvider(`${providerURN}::${providerID}`);
                    }
                    if (result.opts.version !== undefined) {
                        opts.setVersion(result.opts.version);
                    }

                    response.setOptions(opts);
                }
            }
            return response;
        };
        const tryCb = async (bytes: Uint8Array): Promise<jspb.Message> => {
            try {
                return await cb(bytes);
            } catch (e) {
                throw new Error(`transform failed: ${e}`);
            }
        };
        const uuid = randomUUID();
        this._callbacks.set(uuid, tryCb);
        const req = new Callback();
        req.setToken(uuid);
        req.setTarget(await this._target);
        return req;
    }

    registerStackInvokeTransform(transform: InvokeTransform): void {
        this._pendingRegistrations++;

        this.registerStackInvokeTransformAsync(transform)
            .then(
                (req) => {
                    return new Promise((resolve, reject) => {
                        this._monitor.registerStackInvokeTransform(req, (err, _) => {
                            if (err !== null) {
                                // Remove this from the list of callbacks given we didn't manage to actually register it.
                                this._callbacks.delete(req.getToken());
                                reject();
                            } else {
                                resolve();
                            }
                        });
                    });
                },
                (err) => log.error(`failed to register stack transform: ${err}`),
            )
            .finally(() => {
                this._pendingRegistrations--;
                if (this._pendingRegistrations === 0) {
                    const queue = this._awaitQueue;
                    this._awaitQueue = [];
                    for (const waiter of queue) {
                        waiter();
                    }
                }
            });
    }

    async registerResourceHook(hook: ResourceHook): Promise<void> {
        const cb = async (bytes: Uint8Array): Promise<jspb.Message> => {
            try {
                const request = resproto.ResourceHookRequest.deserializeBinary(bytes);
                const newInputs = request.getNewInputs();
                const oldInputs = request.getOldInputs();
                const newOutputs = request.getNewOutputs();
                const oldOutputs = request.getOldOutputs();
                await hook.callback({
                    urn: request.getUrn(),
                    id: request.getId(),
                    name: request.getName(),
                    type: request.getType(),
                    newInputs: newInputs ? deserializeProperties(newInputs, true /*keepUnknowns */) : undefined,
                    oldInputs: oldInputs ? deserializeProperties(oldInputs, true /*keepUnknowns */) : undefined,
                    newOutputs: newOutputs ? deserializeProperties(newOutputs, true /*keepUnknowns */) : undefined,
                    oldOutputs: oldOutputs ? deserializeProperties(oldOutputs, true /*keepUnknowns */) : undefined,
                });
            } catch (error) {
                const response = new resproto.ResourceHookResponse();
                response.setError(error.message);
                return response;
            }
            return new resproto.ResourceHookResponse();
        };

        const uuid = randomUUID();
        this._callbacks.set(uuid, cb);
        const callback = new Callback();
        callback.setToken(uuid);
        callback.setTarget(await this._target);

        const req = new resproto.RegisterResourceHookRequest();
        req.setCallback(callback);
        req.setName(hook.name);
        req.setOnDryRun(hook.opts?.onDryRun ?? false);

        const done = rpcKeepAlive();
        return debuggablePromise(
            new Promise((resolve, reject) => {
                this._monitor.registerResourceHook(req, (err, _) => {
                    if (err !== null) {
                        // Remove this from the list of callbacks given we didn't manage to actually register it.
                        this._callbacks.delete(uuid);
                        reject(err);
                    } else {
                        resolve();
                    }
                    done();
                });
            }),
            `resourceHook:${hook.name}`,
        );
    }
}

File: sdk/python/lib/pulumi/runtime/_callbacks.py
# Copyright 2016-2024, Pulumi Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from typing import (
    TYPE_CHECKING,
    Awaitable,
    Callable,
    Dict,
    List,
    Mapping,
    Union,
    cast,
    Optional,
)
import uuid

import grpc
from grpc import aio

from google.protobuf.message import Message

from .. import log
from .proto import (
    alias_pb2,
    callback_pb2,
    callback_pb2_grpc,
    resource_pb2,
    resource_pb2_grpc,
)
from .rpc import deserialize_properties, serialize_properties
from ..invoke import InvokeOptions, InvokeTransform

if TYPE_CHECKING:
    from ..resource import Alias, ResourceOptions, ResourceTransform


# _MAX_RPC_MESSAGE_SIZE raises the gRPC Max Message size from `4194304` (4mb) to `419430400` (400mb)
_MAX_RPC_MESSAGE_SIZE = 1024 * 1024 * 400
_GRPC_CHANNEL_OPTIONS = [("grpc.max_receive_message_length", _MAX_RPC_MESSAGE_SIZE)]


_CallbackFunction = Callable[[bytes], Awaitable[Message]]


class _CallbackServicer(callback_pb2_grpc.CallbacksServicer):
    _servicers: List[_CallbackServicer] = []

    _callbacks: Dict[str, _CallbackFunction]
    _monitor: resource_pb2_grpc.ResourceMonitorStub
    _server: aio.Server
    _target: str

    _transforms: Dict[Union[ResourceTransform, InvokeTransform], str]

    def __init__(self, monitor: resource_pb2_grpc.ResourceMonitorStub):
        log.debug("Creating CallbackServicer")
        _CallbackServicer._servicers.append(self)
        self._callbacks = {}
        self._transforms = {}
        self._monitor = monitor
        self._server = aio.server(options=_GRPC_CHANNEL_OPTIONS)
        callback_pb2_grpc.add_CallbacksServicer_to_server(self, self._server)
        port = self._server.add_insecure_port(address="127.0.0.1:0")
        self._target = f"127.0.0.1:{port}"

    async def serve(self):
        await self._server.start()

    @classmethod
    async def shutdown(cls):
        for servicer in cls._servicers:
            await servicer._server.wait_for_termination(timeout=0)

    # aio handles this being async but the pyi typings don't expect it.
    async def Invoke(
        self, request: callback_pb2.CallbackInvokeRequest, context
    ) -> callback_pb2.CallbackInvokeResponse:
        log.debug(f"Invoke callback {request.token}")
        callback = self._callbacks.get(request.token)
        if callback is None:
            context.abort(
                grpc.StatusCode.INVALID_ARGUMENT,
                f"Callback with token {request.token} not found!",
            )
            raise Exception("Callback not found!")

        response = await callback(request.request)
        return callback_pb2.CallbackInvokeResponse(
            response=response.SerializeToString()
        )

    def register_transform(self, transform: ResourceTransform) -> callback_pb2.Callback:
        # If this transform function has already been registered, return it.
        token = self._transforms.get(transform)
        if token is not None:
            return callback_pb2.Callback(token=token, target=self._target)

        from ..resource import (
            ResourceTransformArgs,
            ResourceTransformResult,
        )

        async def cb(s: bytes) -> Message:
            request: resource_pb2.TransformRequest = (
                resource_pb2.TransformRequest.FromString(s)
            )

            args = ResourceTransformArgs(
                custom=request.custom,
                type_=request.type,
                name=request.name,
                props=deserialize_properties(request.properties),
                opts=self._resource_options(request),
            )

            maybeAwaitable = transform(args)
            result: Optional[ResourceTransformResult] = None
            if isinstance(maybeAwaitable, Awaitable):
                result = await maybeAwaitable
            else:
                result = maybeAwaitable

            if result is None:
                return resource_pb2.TransformResponse(
                    properties=request.properties,
                    options=request.options,
                )

            result_props = await serialize_properties(result.props, {})

            result_opts = (
                await self._transformation_resource_options(result.opts)
                if result.opts is not None
                else None
            )

            return resource_pb2.TransformResponse(
                properties=result_props,
                options=result_opts,
            )

        token = str(uuid.uuid4())
        self._callbacks[token] = cb
        self._transforms[transform] = token
        return callback_pb2.Callback(
            token=token,
            target=self._target,
        )

    def register_stack_transform(self, transform: ResourceTransform):
        callback = self.register_transform(transform)
        try:
            self._monitor.RegisterStackTransform(callback)
        except:
            # Remove the transform since we didn't manage to actually register it.
            self._transforms.pop(transform)
            self._callbacks.pop(callback.token)
            raise

    def do_register_invoke_transform(
        self, transform: InvokeTransform
    ) -> callback_pb2.Callback:
        # If this transform function has already been registered, return it.
        token = self._transforms.get(transform)
        if token is not None:
            return callback_pb2.Callback(token=token, target=self._target)

        from ..invoke import (
            InvokeTransformArgs,
            InvokeTransformResult,
        )

        async def cb(s: bytes) -> Message:
            request: resource_pb2.TransformInvokeRequest = (
                resource_pb2.TransformInvokeRequest.FromString(s)
            )

            args = InvokeTransformArgs(
                token=request.token,
                args=deserialize_properties(request.args),
                opts=self._invoke_options(request),
            )

            maybeAwaitable = transform(args)
            result: Optional[InvokeTransformResult] = None
            if isinstance(maybeAwaitable, Awaitable):
                result = await maybeAwaitable
            else:
                result = maybeAwaitable

            if result is None:
                return resource_pb2.TransformInvokeResponse(
                    args=request.args,
                    options=request.options,
                )

            result_args = await serialize_properties(result.args, {})

            result_opts = (
                await self._transformation_invoke_options(result.opts)
                if result.opts is not None
                else None
            )

            return resource_pb2.TransformInvokeResponse(
                args=result_args,
                options=result_opts,
            )

        token = str(uuid.uuid4())
        self._callbacks[token] = cb
        self._transforms[transform] = token
        return callback_pb2.Callback(
            token=token,
            target=self._target,
        )

    def register_invoke_transform(self, transform: InvokeTransform):
        callback = self.do_register_invoke_transform(transform)
        try:
            self._monitor.RegisterStackInvokeTransform(callback)
        except:
            # Remove the transform since we didn't manage to actually register it.
            self._transforms.pop(transform)
            self._callbacks.pop(callback.token)
            raise

    def _resource_options(
        self, request: resource_pb2.TransformRequest
    ) -> ResourceOptions:
        from ..resource import (
            CustomTimeouts,
            DependencyProviderResource,
            DependencyResource,
            ResourceOptions,
        )

        opts = (
            request.options
            if request.HasField("options")
            else resource_pb2.TransformResourceOptions()
        )

        ropts = ResourceOptions()

        if opts.HasField("delete_before_replace"):
            ropts.delete_before_replace = opts.delete_before_replace

        additional_secret_outputs = list(opts.additional_secret_outputs)
        if additional_secret_outputs:
            ropts.additional_secret_outputs = additional_secret_outputs

        providers = {
            k: DependencyProviderResource(v) for k, v in opts.providers.items()
        }
        if providers:
            ropts.providers = providers

        aliases = [self._alias(a) for a in opts.aliases]
        if aliases:
            ropts.aliases = aliases

        if opts.HasField("custom_timeouts"):
            custom_timeouts = CustomTimeouts()
            if opts.custom_timeouts.create:
                custom_timeouts.create = opts.custom_timeouts.create
            if opts.custom_timeouts.update:
                custom_timeouts.update = opts.custom_timeouts.update
            if opts.custom_timeouts.delete:
                custom_timeouts.delete = opts.custom_timeouts.delete
            ropts.custom_timeouts = custom_timeouts

        if opts.deleted_with:
            ropts.deleted_with = DependencyResource(opts.deleted_with)

        depends_on = [DependencyResource(d) for d in opts.depends_on]
        if depends_on:
            ropts.depends_on = depends_on

        ignore_changes = list(opts.ignore_changes)
        if ignore_changes:
            ropts.ignore_changes = ignore_changes

        if request.parent:
            ropts.parent = DependencyResource(request.parent)

        if opts.plugin_download_url:
            ropts.plugin_download_url = opts.plugin_download_url

        if opts.HasField("protect"):
            ropts.protect = opts.protect

        if opts.provider:
            ropts.provider = DependencyProviderResource(opts.provider)

        replace_on_changes = list(opts.replace_on_changes)
        if replace_on_changes:
            ropts.replace_on_changes = replace_on_changes

        if opts.retain_on_delete:
            ropts.retain_on_delete = opts.retain_on_delete

        if opts.version:
            ropts.version = opts.version

        return ropts

    def _alias(self, alias: alias_pb2.Alias) -> Union[str, Alias]:
        if alias.HasField("urn"):
            return alias.urn

        from ..resource import Alias

        a = Alias()
        if alias.spec.name:
            a.name = alias.spec.name
        if alias.spec.type:
            a.type_ = alias.spec.type
        if alias.spec.project:
            a.project = alias.spec.project
        if alias.spec.stack:
            a.stack = alias.spec.stack
        if alias.spec.parentUrn:
            a.parent = alias.spec.parentUrn
        elif alias.spec.noParent:
            a.parent = None
        return a

    async def _transformation_resource_options(
        self, opts: ResourceOptions
    ) -> resource_pb2.TransformResourceOptions:
        from ._depends_on import (
            _resolve_depends_on_urns,
        )
        from .resource import (
            _create_custom_timeouts,
            _create_provider_ref,
            create_alias_spec,
        )
        from ..resource import (
            Alias,
            ProviderResource,
            _collapse_providers,
        )
        from ..output import Output

        aliases: List[alias_pb2.Alias] = []
        if opts.aliases:
            for alias in opts.aliases:
                resolved = await Output.from_input(alias).future()
                if resolved is None:
                    continue
                if isinstance(resolved, str):
                    aliases.append(alias_pb2.Alias(urn=resolved))
                else:
                    spec = await create_alias_spec(cast(Alias, resolved))
                    aliases.append(alias_pb2.Alias(spec=spec))

        custom_timeouts = None
        if opts.custom_timeouts is not None:
            custom_timeouts = _create_custom_timeouts(opts.custom_timeouts)

        depends_on = await _resolve_depends_on_urns(opts._depends_on_list())

        ignore_changes = None
        if opts.ignore_changes:
            ignore_changes = list(opts.ignore_changes)

        replace_on_changes = None
        if opts.replace_on_changes:
            replace_on_changes = list(opts.replace_on_changes)

        additional_secret_outputs = None
        if opts.additional_secret_outputs:
            additional_secret_outputs = list(opts.additional_secret_outputs)

        result = resource_pb2.TransformResourceOptions(
            aliases=aliases or None,
            custom_timeouts=custom_timeouts,
            depends_on=depends_on or None,
            ignore_changes=ignore_changes,
            replace_on_changes=replace_on_changes,
            additional_secret_outputs=additional_secret_outputs,
        )

        if opts.deleted_with is not None:
            result.deleted_with = cast(str, await opts.deleted_with.urn.future())
        if opts.plugin_download_url:
            result.plugin_download_url = opts.plugin_download_url
        if opts.protect is not None:
            result.protect = opts.protect
        if opts.retain_on_delete:
            result.retain_on_delete = opts.retain_on_delete
        if opts.version:
            result.version = opts.version
        if opts.provider is not None:
            result.provider = await _create_provider_ref(opts.provider)
        if opts.delete_before_replace:
            result.delete_before_replace = opts.delete_before_replace
        if opts.providers:
            _collapse_providers(opts)
            providers = cast(Mapping[str, ProviderResource], opts.providers)
            result.providers.update(
                {k: await _create_provider_ref(v) for k, v in providers.items()}
            )

        return result

    def _invoke_options(
        self, request: resource_pb2.TransformInvokeRequest
    ) -> InvokeOptions:
        from ..resource import (
            DependencyProviderResource,
        )

        opts = (
            request.options
            if request.HasField("options")
            else resource_pb2.TransformInvokeOptions()
        )

        ropts = InvokeOptions()

        if opts.plugin_download_url:
            ropts.plugin_download_url = opts.plugin_download_url

        if opts.provider:
            ropts.provider = DependencyProviderResource(opts.provider)

        if opts.version:
            ropts.version = opts.version

        return ropts

    async def _transformation_invoke_options(
        self, opts: InvokeOptions
    ) -> resource_pb2.TransformInvokeOptions:
        from .resource import (
            _create_provider_ref,
        )

        result = resource_pb2.TransformInvokeOptions()

        if opts.plugin_download_url:
            result.plugin_download_url = opts.plugin_download_url
        if opts.version:
            result.version = opts.version
        if opts.provider is not None:
            result.provider = await _create_provider_ref(opts.provider)

        return result
